In [7]:
import pandas as pd

df = pd.read_json("df_normalizado.json")  # reemplaza con tu nombre de archivo
df.head()


customerID Churn customer.gender  customer.SeniorCitizen customer.Partner  \
0  0002-ORFBO    No          Female                       0              Yes   
1  0003-MKNFE    No            Male                       0               No   
2  0004-TLHLJ   Yes            Male                       0               No   
3  0011-IGKFF   Yes            Male                       1              Yes   
4  0013-EXCHZ   Yes          Female                       1              Yes   

  customer.Dependents  customer.tenure phone.PhoneService phone.MultipleLines  \
0                 Yes                9                Yes                  No   
1                  No                9                Yes                 Yes   
2                  No                4                Yes                  No   
3                  No               13                Yes                  No   
4                  No                3                Yes                  No   

  internet.InternetService  ... internet.DeviceProtection  \
0                      DSL  ...                        No   
1                      DSL  ...                        No   
2              Fiber optic  ...                       Yes   
3              Fiber optic  ...                       Yes   
4              Fiber optic  ...                        No   

  internet.TechSupport internet.StreamingTV internet.StreamingMovies  \
0                  Yes                  Yes                       No   
1                   No                   No                      Yes   
2                   No                   No                       No   
3                   No                  Yes                      Yes   
4                  Yes                  Yes                       No   

  account.Contract account.PaperlessBilling account.PaymentMethod  \
0         One year                      Yes          Mailed check   
1   Month-to-month                       No          Mailed check   
2   Month-to-month                      Yes      Electronic check   
3   Month-to-month                      Yes      Electronic check   
4   Month-to-month                      Yes          Mailed check   

  account.Charges.Monthly account.Charges.Total  Cuentas_Diarias  
0                    65.6                593.30         2.186667  
1                    59.9                542.40         1.996667  
2                    73.9                280.85         2.463333  
3                    98.0               1237.85         3.266667  
4                    83.9                267.40         2.796667  

[5 rows x 22 columns]

In [6]:
#df = df.drop(df.columns[0], axis=1)
df = df.drop('customerID', axis=1)


NameError: name 'df' is not defined

In [ ]:
print(df.columns.tolist())

# Mostrar las primeras 5 filas
print(df.head())


In [ ]:
print(df.shape)
df.head()


In [ ]:
for i, col in enumerate(df.columns):
    print(f"{i}: {col}")


In [ ]:
# Todas las columnas con su tipo
print(df.dtypes)

# Solo columnas categóricas (tipo object o category)
cat_cols = df.select_dtypes(include=['object', 'category']).columns
print("Columnas categóricas:", list(cat_cols))


In [8]:
# Para cada columna, mostrar la cantidad de valores únicos
unique_counts = df.nunique()
print(unique_counts)

customerID                   7043
Churn                           2
customer.gender                 2
customer.SeniorCitizen          2
customer.Partner                2
customer.Dependents             2
customer.tenure                73
phone.PhoneService              2
phone.MultipleLines             3
internet.InternetService        3
internet.OnlineSecurity         3
internet.OnlineBackup           3
internet.DeviceProtection       3
internet.TechSupport            3
internet.StreamingTV            3
internet.StreamingMovies        3
account.Contract                3
account.PaperlessBilling        2
account.PaymentMethod           4
account.Charges.Monthly      1585
account.Charges.Total        6530
Cuentas_Diarias              1585
dtype: int64


In [9]:
cat_cols = df.select_dtypes(include=['object']).columns

for col in cat_cols:
    print(f"Opciones en la columna '{col}':")
    print(df[col].unique())
    print("-" * 40)


Opciones en la columna 'customerID':
['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
----------------------------------------
Opciones en la columna 'Churn':
['No' 'Yes']
----------------------------------------
Opciones en la columna 'customer.gender':
['Female' 'Male']
----------------------------------------
Opciones en la columna 'customer.Partner':
['Yes' 'No']
----------------------------------------
Opciones en la columna 'customer.Dependents':
['Yes' 'No']
----------------------------------------
Opciones en la columna 'phone.PhoneService':
['Yes' 'No']
----------------------------------------
Opciones en la columna 'phone.MultipleLines':
['No' 'Yes' 'No phone service']
----------------------------------------
Opciones en la columna 'internet.InternetService':
['DSL' 'Fiber optic' 'No']
----------------------------------------
Opciones en la columna 'internet.OnlineSecurity':
['No' 'Yes' 'No internet service']
-------------------------------

In [13]:
df.to_csv("df_sin_customerID.csv", index=False, encoding="utf-8-sig")
print(df.head(10))

  Churn customer.gender  customer.SeniorCitizen customer.Partner  \
0    No          Female                       0              Yes   
1    No            Male                       0               No   
2   Yes            Male                       0               No   
3   Yes            Male                       1              Yes   
4   Yes          Female                       1              Yes   
5    No          Female                       0               No   
6    No          Female                       1              Yes   
7    No            Male                       0              Yes   
8    No          Female                       1               No   
9    No          Female                       0              Yes   

  customer.Dependents  customer.tenure phone.PhoneService phone.MultipleLines  \
0                 Yes                9                Yes                  No   
1                  No                9                Yes                 Yes   
2       

In [15]:
for col in df.columns:
    print(col)

Churn
customer.gender
customer.SeniorCitizen
customer.Partner
customer.Dependents
customer.tenure
phone.PhoneService
phone.MultipleLines
internet.InternetService
internet.OnlineSecurity
internet.OnlineBackup
internet.DeviceProtection
internet.TechSupport
internet.StreamingTV
internet.StreamingMovies
account.Contract
account.PaperlessBilling
account.PaymentMethod
account.Charges.Monthly
account.Charges.Total
Cuentas_Diarias


In [16]:
# Lista columnas que solo tienen 'Yes' y 'No'
cols_yes_no = [col for col in df.columns if set(df[col].dropna().unique()) == {'Yes', 'No'}]

print("Columnas con solo valores 'Yes' y 'No':")
print(cols_yes_no)

Columnas con solo valores 'Yes' y 'No':
['Churn', 'customer.Partner', 'customer.Dependents', 'phone.PhoneService', 'account.PaperlessBilling']


In [17]:
# Reemplazar Yes=1, No=0 en esas columnas
for col in cols_yes_no:
    df[col] = df[col].map({'Yes': 1, 'No': 0})

# Verificamos cambios
print(df[cols_yes_no].head())

   Churn  customer.Partner  customer.Dependents  phone.PhoneService  \
0      0                 1                    1                   1   
1      0                 0                    0                   1   
2      1                 0                    0                   1   
3      1                 1                    0                   1   
4      1                 1                    0                   1   

   account.PaperlessBilling  
0                         1  
1                         0  
2                         1  
3                         1  
4                         1  


In [18]:
# Encontrar columnas con exactamente 3 valores únicos
cols_tres_valores = [col for col in df.columns if df[col].nunique() == 3]

print("Columnas con exactamente 3 valores únicos:")
for col in cols_tres_valores:
    print(col)


Columnas con exactamente 3 valores únicos:
phone.MultipleLines
internet.InternetService
internet.OnlineSecurity
internet.OnlineBackup
internet.DeviceProtection
internet.TechSupport
internet.StreamingTV
internet.StreamingMovies
account.Contract


In [19]:
for col in df.columns:
    unique_vals = df[col].dropna().unique()
    if len(unique_vals) == 3:
        print(f"Columna '{col}' tiene 3 valores únicos:")
        print(unique_vals)
        print("-" * 40)

Columna 'phone.MultipleLines' tiene 3 valores únicos:
['No' 'Yes' 'No phone service']
----------------------------------------
Columna 'internet.InternetService' tiene 3 valores únicos:
['DSL' 'Fiber optic' 'No']
----------------------------------------
Columna 'internet.OnlineSecurity' tiene 3 valores únicos:
['No' 'Yes' 'No internet service']
----------------------------------------
Columna 'internet.OnlineBackup' tiene 3 valores únicos:
['Yes' 'No' 'No internet service']
----------------------------------------
Columna 'internet.DeviceProtection' tiene 3 valores únicos:
['No' 'Yes' 'No internet service']
----------------------------------------
Columna 'internet.TechSupport' tiene 3 valores únicos:
['Yes' 'No' 'No internet service']
----------------------------------------
Columna 'internet.StreamingTV' tiene 3 valores únicos:
['Yes' 'No' 'No internet service']
----------------------------------------
Columna 'internet.StreamingMovies' tiene 3 valores únicos:
['No' 'Yes' 'No interne

In [20]:
no_service_count = df['phone.MultipleLines'].value_counts().get('No phone service', 0)
print(f"Clientes sin servicio telefónico: {no_service_count}")

Clientes sin servicio telefónico: 682


In [21]:
conteo = df['phone.MultipleLines'].value_counts()
print(conteo)

phone.MultipleLines
No                  3390
Yes                 2971
No phone service     682
Name: count, dtype: int64


Perfecto, esos números te dicen que:

3390 clientes tienen teléfono, pero una sola línea (No)
2971 clientes tienen múltiples líneas (Yes)
682 clientes no tienen servicio telefónico (No phone service)

Qué puedes hacer con esta info:
La categoría 'No phone service' es relevante porque representa casi el 11% de tus clientes (682 de 7043 aprox).

Por eso, mantenerla como categoría separada en tu análisis o modelo es muy recomendable.


Cómo hacerlo para regresión:
Usa One-Hot Encoding para transformar la columna phone.MultipleLines en variables binarias:

phone.MultipleLines_No

phone.MultipleLines_Yes

phone.MultipleLines_No phone service

Esto permite que el modelo aprenda efectos diferentes para cada grupo, sin asumir orden ni mezcla.

Ventajas de mantener la categoría separada:
El modelo captura diferencias reales entre tener múltiples líneas, tener una sola línea y no tener servicio.

Evitas perder información valiosa que podría afectar la predicción.

Consideraciones:
Si usas pd.get_dummies(..., drop_first=True), una categoría (la base) se elimina para evitar multicolinealidad.

Si quieres mantener explícitamente todas las categorías, no uses drop_first=True o usa otro método como OneHotEncoder de scikit-learn con drop=None.

Asegúrate de interpretar los coeficientes en regresión teniendo en cuenta la categoría base.

Mantendre las tres opciones.

¡Perfecto! Esa es la mejor estrategia.

Para mantener la categoría 'No phone service' aparte y evitar multicolinealidad, lo que puedes hacer es:

Aplicar One-Hot Encoding a toda la columna phone.MultipleLines sin usar drop_first=True, para que todas las categorías tengan su propia columna.

Luego, para evitar multicolinealidad, cuando uses modelos lineales (regresión, regresión logística), aplica regularización (como Lasso o Ridge) o bien elimina manualmente una columna base (puedes elegir cuál eliminar según conveniencia).

In [22]:
# Aplicar One-Hot Encoding a 'phone.MultipleLines' sin eliminar ninguna categoría
df_encoded = pd.get_dummies(df, columns=['phone.MultipleLines'], drop_first=False)

# Mostrar las nuevas columnas creadas para esa variable
print(df_encoded.filter(like='phone.MultipleLines').head())

# (Opcional) Para evitar multicolinealidad en modelos lineales, eliminar una columna dummy
df_encoded = df_encoded.drop('phone.MultipleLines_No', axis=1)

# Mostrar columnas después de la eliminación
print(df_encoded.filter(like='phone.MultipleLines').head())

   phone.MultipleLines_No  phone.MultipleLines_No phone service  \
0                    True                                 False   
1                   False                                 False   
2                    True                                 False   
3                    True                                 False   
4                    True                                 False   

   phone.MultipleLines_Yes  
0                    False  
1                     True  
2                    False  
3                    False  
4                    False  
   phone.MultipleLines_No phone service  phone.MultipleLines_Yes
0                                 False                    False
1                                 False                     True
2                                 False                    False
3                                 False                    False
4                                 False                    False


In [23]:
print(df['phone.MultipleLines'].unique())


['No' 'Yes' 'No phone service']


In [24]:
# Reemplazar espacios por guiones bajos en todas las columnas de tipo object
cat_cols = df.select_dtypes(include=['object']).columns
for col in cat_cols:
    df[col] = df[col].str.replace(' ', '_')

# Ahora aplicas get_dummies sin problemas
df_encoded = pd.get_dummies(df, drop_first=False)

# Verificas columnas relacionadas con phone.MultipleLines
print(df_encoded.filter(like='phone.MultipleLines').head())


   phone.MultipleLines_No  phone.MultipleLines_No_phone_service  \
0                    True                                 False   
1                   False                                 False   
2                    True                                 False   
3                    True                                 False   
4                    True                                 False   

   phone.MultipleLines_Yes  
0                    False  
1                     True  
2                    False  
3                    False  
4                    False  


Perfecto, vamos a trabajar con la columna internet.InternetService.

Esta columna tiene tres categorías:"DSL", "Fiber optic", "No"

En machine learning, lo más común es codificar variables categóricas de este tipo usando One-Hot Encoding, porque son valores nominales (no existe un orden lógico entre ellos).

Aquí depende del objetivo de tu análisis o modelo:

Si estás haciendo un análisis descriptivo (por ejemplo, para ver cuántos clientes usan cada tipo de servicio), deja los 3 valores tal cual:

"DSL"
"Fiber optic"
"No" (clientes sin servicio de Internet).
Si estás preparando datos para un modelo de machine learning, a veces conviene:

Mantener los 3 valores y codificarlos (one-hot encoding o label encoding).

En la práctica, si dejas los 3 y haces one-hot encoding, el modelo podrá aprender la diferencia sin problema.

In [28]:
# 1️⃣ Crear un DataFrame de ejemplo (puedes saltar este paso si ya tienes el tuyo)
data = {
    'phone.MultipleLines': ['No', 'Yes', 'No phone service', 'Yes', 'No']
}
df = pd.DataFrame(data)

# 2️⃣ Mostrar valores únicos antes de reemplazar
print("Valores únicos originales:")
print(df['phone.MultipleLines'].unique())

# 3️⃣ Reemplazar los valores
df['phone.MultipleLines'] = df['phone.MultipleLines'].replace({
    'No': 'No tiene múltiples líneas',
    'Yes': 'Sí tiene múltiples líneas',
    'No phone service': 'No tiene servicio telefónico'
})

# 4️⃣ Mostrar valores únicos después de reemplazar
print("\nValores únicos después de reemplazar:")
print(df['phone.MultipleLines'].unique())

# 5️⃣ Mostrar DataFrame resultante
df

Valores únicos originales:
['No' 'Yes' 'No phone service']

Valores únicos después de reemplazar:
['No tiene múltiples líneas' 'Sí tiene múltiples líneas'
 'No tiene servicio telefónico']


phone.MultipleLines
0     No tiene múltiples líneas
1     Sí tiene múltiples líneas
2  No tiene servicio telefónico
3     Sí tiene múltiples líneas
4     No tiene múltiples líneas

In [29]:
df['phone.MultipleLines'].head(3)


0       No tiene múltiples líneas
1       Sí tiene múltiples líneas
2    No tiene servicio telefónico
Name: phone.MultipleLines, dtype: object

En la columna internet.OnlineSecurity con valores:

Yes: significa que el cliente tiene activado el servicio de seguridad en línea.

No: significa que el cliente no tiene activado el servicio de seguridad en línea, pero sí tiene servicio de internet (DSL o Fiber).

No internet service: significa que el cliente no tiene servicio de internet contratado, por lo tanto la pregunta sobre seguridad en línea no aplica.

Resumen:
Valor	Significado
Yes	Cliente tiene activado Online Security
No	Cliente tiene internet pero no activó la seguridad
No internet service	Cliente no tiene internet, por eso no aplica

In [31]:
# Ejemplo DataFrame
df = pd.DataFrame({
    'internet.OnlineSecurity': ['No', 'Yes', 'No internet service', 'Yes', 'No']
})

# Reemplazo valores con 0,1,2
df['internet.OnlineSecurity'] = df['internet.OnlineSecurity'].replace({
    'Yes': 1,
    'No': 0,
    'No internet service': 2
})

print(df)

   internet.OnlineSecurity
0                        0
1                        1
2                        2
3                        1
4                        0


/tmp/ipython-input-1286931429.py:7: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['internet.OnlineSecurity'] = df['internet.OnlineSecurity'].replace({


In [33]:
df['internet.OnlineSecurity'] = df['internet.OnlineSecurity'].replace({
    'Yes': 1,
    'No': 0,
    'No internet service': 2
}).astype(int)


In [34]:
df['internet.OnlineSecurity'].head(3)

0    0
1    1
2    2
Name: internet.OnlineSecurity, dtype: int64

Porque la columna internet.InternetService tiene 3 categorías distintas:

DSL

Fiber optic

No

Cuando aplicas One-Hot Encoding, pandas crea una columna binaria para cada categoría, con valor 1 si el registro tiene esa categoría, y 0 en caso contrario.

Entonces, para 3 categorías, obtienes 3 columnas:

internet.InternetService_DSL	internet.InternetService_Fiber_optic	internet.InternetService_No
1	0	0
0	1	0
0	0	1

In [37]:
import pandas as pd

# Supongamos df original
df = pd.DataFrame({
    'internet.InternetService': ['DSL', 'Fiber optic', 'No', 'DSL', 'Fiber optic']
})

print("Antes de One-Hot Encoding:")
print(df.head())

# Aplico one-hot encoding y actualizo df
df = pd.get_dummies(df, columns=['internet.InternetService'], drop_first=False)

print("\nDespués de One-Hot Encoding:")
print(df.head())

Antes de One-Hot Encoding:
  internet.InternetService
0                      DSL
1              Fiber optic
2                       No
3                      DSL
4              Fiber optic

Después de One-Hot Encoding:
   internet.InternetService_DSL  internet.InternetService_Fiber optic  \
0                          True                                 False   
1                         False                                  True   
2                         False                                 False   
3                          True                                 False   
4                         False                                  True   

   internet.InternetService_No  
0                        False  
1                        False  
2                         True  
3                        False  
4                        False  


In [40]:
print(df.filter(like='internet.InternetService').head(3))

   internet.InternetService_DSL  internet.InternetService_Fiber optic  \
0                          True                                 False   
1                         False                                  True   
2                         False                                 False   

   internet.InternetService_No  
0                        False  
1                        False  
2                         True  


In [41]:
print(df.columns)

Index(['internet.InternetService_DSL', 'internet.InternetService_Fiber optic',
       'internet.InternetService_No'],
      dtype='object')


In [42]:
print(df.dtypes[['internet.InternetService_DSL', 'internet.InternetService_Fiber optic', 'internet.InternetService_No']])


internet.InternetService_DSL            bool
internet.InternetService_Fiber optic    bool
internet.InternetService_No             bool
dtype: object


In [43]:
# Selecciona solo las columnas dummies que creaste (ejemplo con las 3 columnas)
cols = ['internet.InternetService_DSL', 'internet.InternetService_Fiber optic', 'internet.InternetService_No']

# Convierte de bool a int
df[cols] = df[cols].astype(int)

# Ahora verás 0 y 1 en vez de False y True
print(df[cols].head())

   internet.InternetService_DSL  internet.InternetService_Fiber optic  \
0                             1                                     0   
1                             0                                     1   
2                             0                                     0   
3                             1                                     0   
4                             0                                     1   

   internet.InternetService_No  
0                            0  
1                            0  
2                            1  
3                            0  
4                            0  


In [45]:
print(df.columns.tolist())


['internet.InternetService_DSL', 'internet.InternetService_Fiber optic', 'internet.InternetService_No']


In [46]:
import pandas as pd

# Lista completa de columnas que quieres procesar (ejemplo)
cols_3vals_all = [
    'internet.OnlineBackup',
    'internet.DeviceProtection',
    'internet.TechSupport',
    'internet.StreamingTV',
    'internet.StreamingMovies'
]

cols_binarias_all = [
    'customer.Partner',
    'customer.Dependents',
    'phone.PhoneService',
    'account.PaperlessBilling'
]

cols_onehot_all = [
    'account.Contract',
    'account.PaymentMethod',
    'phone.MultipleLines',
    'internet.InternetService'
]

# Filtrar solo las columnas que existen en df
cols_3vals = [col for col in cols_3vals_all if col in df.columns]
cols_binarias = [col for col in cols_binarias_all if col in df.columns]
cols_onehot = [col for col in cols_onehot_all if col in df.columns]

# Aplicar reemplazos a columnas con 3 valores
for col in cols_3vals:
    df[col] = df[col].replace({'Yes': 1, 'No': 0, 'No internet service': 2})

# Aplicar reemplazos a columnas binarias
for col in cols_binarias:
    df[col] = df[col].replace({'Yes': 1, 'No': 0})

# Aplicar one-hot encoding
df = pd.get_dummies(df, columns=cols_onehot, drop_first=False)

# Convertir booleanos a enteros
bool_cols = df.select_dtypes(include='bool').columns
df[bool_cols] = df[bool_cols].astype(int)

print(df.head())


   internet.InternetService_DSL  internet.InternetService_Fiber optic  \
0                             1                                     0   
1                             0                                     1   
2                             0                                     0   
3                             1                                     0   
4                             0                                     1   

   internet.InternetService_No  
0                            0  
1                            0  
2                            1  
3                            0  
4                            0  


In [47]:
# Mostrar las primeras 10 filas para ver los datos
print(df.head(10))

# Mostrar las columnas actuales y su tipo de dato
print(df.dtypes)

# Mostrar la cantidad de valores únicos por columna para verificar codificaciones
print(df.nunique())


   internet.InternetService_DSL  internet.InternetService_Fiber optic  \
0                             1                                     0   
1                             0                                     1   
2                             0                                     0   
3                             1                                     0   
4                             0                                     1   

   internet.InternetService_No  
0                            0  
1                            0  
2                            1  
3                            0  
4                            0  
internet.InternetService_DSL            int64
internet.InternetService_Fiber optic    int64
internet.InternetService_No             int64
dtype: object
internet.InternetService_DSL            2
internet.InternetService_Fiber optic    2
internet.InternetService_No             2
dtype: int64


In [49]:
# Mostrar columnas que siguen siendo tipo object (categóricas sin transformar)
cols_obj = df.select_dtypes(include='object').columns.tolist()
print("Columnas aún tipo objeto (posibles sin transformar):")
print(cols_obj)


Columnas aún tipo objeto (posibles sin transformar):
[]


In [51]:
print(df.columns.tolist())


['internet.InternetService_DSL', 'internet.InternetService_Fiber optic', 'internet.InternetService_No']


In [52]:
print(df.columns.tolist())


['internet.InternetService_DSL', 'internet.InternetService_Fiber optic', 'internet.InternetService_No']
